### PR TITLE
Add skyline selection

### PR DESCRIPTION
**Pull request by KeYang0923**
No overwrite of original .py scripts. Original .py scripts are renamed with "_old" suffix.

### CODE REVIEW

File: examples_skyline/fair_prep_results.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+from fp.learners import NonTunedLogisticRegression, LogisticRegression, DecisionTree, NonTunedLogisticRegression, NonTunedDecisionTree          
+from fp.post_processors import NoPostProcessing, RejectOptionPostProcessing, EqualOddsPostProcessing, CalibratedEqualOddsPostProcessing
+from fp.pre_processors import NoPreProcessing, DIRemover, Reweighing
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+#creating list of parameters that we will alter to observe variations
+seeds = [0xbeef, 0xcafe, 0xdead, 0xdeadcafe, 0xdeadbeef, 0xbeefcafe, 0xcafebeef, 50, 583, 5278, 100000, 0xefac,0xfeeb, 0xdaed, 0xefacdaed, 0xfeebdead]
+learners = [NonTunedLogisticRegression(), LogisticRegression()]
```
Comment (sscdotopen) [inline examples_skyline/fair_prep_results.py:24]: I think we should remove the non-tuned case, it should not be used at all.

File: examples_skyline/fair_prep_results.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+from fp.scalers import NamedStandardScaler, NamedMinMaxScaler
+from fp.learners import NonTunedLogisticRegression, LogisticRegression, DecisionTree, NonTunedLogisticRegression, NonTunedDecisionTree          
+from fp.post_processors import NoPostProcessing, RejectOptionPostProcessing, EqualOddsPostProcessing, CalibratedEqualOddsPostProcessing
+from fp.pre_processors import NoPreProcessing, DIRemover, Reweighing
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+#creating list of parameters that we will alter to observe variations
+seeds = [0xbeef, 0xcafe, 0xdead, 0xdeadcafe, 0xdeadbeef, 0xbeefcafe, 0xcafebeef, 50, 583, 5278, 100000, 0xefac,0xfeeb, 0xdaed, 0xefacdaed, 0xfeebdead]
```
Comment (sscdotopen) [inline examples_skyline/fair_prep_results.py:23]: Maybe we can also reduce the number of random seeds

File: examples_skyline/helper.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                for file in file_list:
+                    file_path = os.path.join(path,sub_folder, file)
+                    df = pd.read_csv(str(file_path), header=None, names = ['split_type', 'label', 'parameter', 'value'])
+                    test_data = df['split_type'] == 'test'
+                    
+                    flags = file.split("__")
+                    pre_processor = flags[1]
+                    post_processor = flags[2]
+ 
+                    if(len(flags) == 8 and flags[-1] == "OPTIMAL.csv"):
```
Comment (sscdotopen) [inline examples_skyline/helper.py:23]: space after if

File: examples_skyline/helper.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                        learner_optimal.append(True)
+                    else:
+                        learner_optimal.append(False)
+                        
+                    if(post_processor == "no_post_processing"):
```
Comment (sscdotopen) [inline examples_skyline/helper.py:28]: space after if

File: examples_skyline/helper.py (hunk 3 above)
Comment (sscdotopen) [inline examples_skyline/helper.py:23]: please refactor this code into its own method, its really hard to understand what this is doing, so give the method a speaking name etc

File: examples_skyline/helper.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                        learner_label.append(pre_processor)
+                    else:
+                        learner_label.append(post_processor[:17])
```
Comment (sscdotopen) [inline examples_skyline/helper.py:31]: what is this doing?

File: fp/dataset_experiments_old.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
+from fp.experiments import BinaryClassificationExperiment
```
Comment (sscdotopen) [inline fp/dataset_experiments_old.py:1]: why do we need an _old version of this file?

File: fp/experiments.py [hunk 7]
```diff
@@ -59,20 +61,21 @@ def __init__(self,
...
 
 
 
     # --- Helper Methods Begin ------------------------------------------------
 
 
-    def unique_file_name(self, learner, pre_processor, post_processor):
-        return '{}__{}__{}__{}__{}__{}__{}'.format(self.dataset_name,
+    def unique_file_name(self, pre_processor, post_processor, learner):
+        return '{0}__{1}__{2}__{3}__{4}__{5}__{6}'.format(self.dataset_name,
```
Comment (sscdotopen) [inline fp/experiments.py:72]: why do we have these numbers here?

File: fp/experiments.py [hunk 8]
```diff
@@ -280,45 +280,70 @@ def run_single_exp(self, annotated_train_data, validation_data, test_data, scale
...
             optimal result.
         """
+
         results_dir_name = os.path.join(os.path.dirname(os.path.realpath(__file__)), '../{}'.format(self.generate_file_path()))
         results_dir = os.listdir(Path(results_dir_name))
-        accuracies = dict()
-        max_accuracy = 0
 
-        # Fetching the accuracy from the row('val', 'None', 'accuracy') of all the experiment results
+        ##### SKYLINE FORMULA IMPLEMENTATION
```
Comment (sscdotopen) [inline fp/experiments.py:293]: please refactor the skyline code into its own method.

File: fp/experiments.py [hunk 9]
```diff
@@ -280,45 +280,70 @@ def run_single_exp(self, annotated_train_data, validation_data, test_data, scale
...
+                                   'num_generalized_false_positives', 'num_generalized_false_negatives',
+                                   'num_generalized_true_negatives', 'true_positive_rate', 'false_positive_rate',
+                                   'false_negative_rate', 'true_negative_rate', 'generalized_true_positive_rate',
+                                   'generalized_false_positive_rate', 'generalized_false_negative_rate',
+                                   'generalized_true_negative_rate', 'positive_predictive_value',
+                                   'false_discovery_rate', 'false_omission_rate', 'negative_predictive_value',
+                                   'accuracy', 'error_rate', 'num_pred_positives', 'num_pred_negatives',
+                                   'selection_rate']
+         
+        dictionary = {}
```
Comment (sscdotopen) [inline fp/experiments.py:306]: please give this variable a different name so that a reader understand what is purpose is

File: fp/utils.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import pandas as pd
+
+def filter_optimal_results_skyline_order(_df, _order_list):
```
Comment (sscdotopen) [inline fp/utils.py:3]: why do we use underscores here?

File: examples/skyline_selection_results.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+import seaborn as sns
+
+# creating list of parameters that we will alter to observe variations
+seeds = [0xbeef, 0xcafe, 0xdead, 0xdeadcafe]
+learners = [LogisticRegression()]
+
+processors = [(NoPreProcessing(), NoPostProcessing()), (DIRemover(1.0), NoPostProcessing()), (NoPreProcessing(), RejectOptionPostProcessing())]
+
+
+skyline_order = ['accuracy', 'selection_rate', 'false_discovery_rate']
```
Comment (sscdotopen) [inline examples/skyline_selection_results.py:27]: most users might not know what a skyline is, can we give these variables a different name?

File: examples/skyline_selection_results.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    exp.run()
+    return exp.generate_file_path()
+
+def run_exp(seeds, learners, processors, skyline_strategy):
+    '''
+        This is the main driver function that calls the calculate_metrics to give metrices on combinations of various learners, pre and post processing techniques.
+    '''
+    skyline_res_folder = {}
+    for seed in seeds:
+        skyline_res_folder[seed] = calculate_metrics(seed, learners, [x[0] for x in processors], [x[1] for x in processors], skyline_strategy)
```
Comment (sscdotopen) [inline examples/skyline_selection_results.py:52]: the comprehensions for the preprocessors are difficult to read, please just give two lists to the function

File: examples/skyline_selection_results.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+
+def output_scatter_plot(f_name, df, x_col, y_col, hue_col='setting', color_p='Set2'):
+    '''
+        Visualization of the skyline options w.r.t. two metrics (X and Y axis) in the skyline inputs.
+    '''
+    sns.set(style='whitegrid', font_scale=1.5)
+    # add jitters for x to account for ties in the values
+    data = df.copy()
+    noise_para = 100
```
Comment (sscdotopen) [inline examples/skyline_selection_results.py:63]: noise_param

File: examples/skyline_selection_results.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+def get_skyline_candidates(seed_path_map, focus_seed):
+    '''
+        Prepare the skyline candidates data for visualization
+    '''
+    setting_labels = {'reject_option': 'RO', 'diremover-1.0': 'DI1.0',
+                      'no_pre_processing': 'NoPre', 'no_post_processing': 'NoPost',
+                      'DecisionTree': 'DT', 'LogisticRegression': 'LR'}
+    skyline_df = pd.read_csv(seed_path_map[focus_seed] + "skyline_options.csv")
+    # rename the setting of preprocessor (idx 1), learner (idx 5), and postprocessor (idx 6)
+    skyline_df['setting'] = skyline_df['setting'].apply(lambda x: '_'.join([setting_labels[x.split('__')[i].replace('-' + str(focus_seed), '')] for i in range(len(x.split('__'))) if i in [1, 5, 6]]))
```
Comment (sscdotopen) [inline examples/skyline_selection_results.py:90]: this line is extremely hard to read, whats happening here?

File: fp/utils.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        if item in higher_is_better:
+            order.append(False)
+        else:
+            order.append(True)
+    _df = _df.sort_values(_order_list, ascending=order)
+
+    return _df.values[0]
+
+
+def filter_optimal_results_skyline_formula(_df, _formula):
```
Comment (sscdotopen) [inline fp/utils.py:25]: please give names to variables that explain what data they contain, using just df and _df is very confusing

Comment (KeYang0923) [reply fp/utils.py]: Fixed by deleting it. This file should be deleted for the refactored code.

File: examples/skyline_selection_results.py (hunk 14 above)
Comment (KeYang0923) [reply examples/skyline_selection_results.py]: Updated to three lines for readability.

File: examples/skyline_selection_results.py (hunk 13 above)
Comment (KeYang0923) [reply examples/skyline_selection_results.py]: Fixed.

File: examples/skyline_selection_results.py (hunk 12 above)
Comment (KeYang0923) [reply examples/skyline_selection_results.py]: Fixed.

File: examples/skyline_selection_results.py (hunk 11 above)
Comment (KeYang0923) [reply examples/skyline_selection_results.py]: Fixed by renaming the variables and adding explanations for readability.

**Pull request merged**